File: python/system_design/ex3_rate_limiting.py
"""Exercise 3: Rate Limiting"""
import time
from collections import deque

class TokenBucketRateLimiter:
    def __init__(self, capacity: int, refill_rate: float):
        # TODO: Implement token bucket
        raise NotImplementedError()

    def allow_request(self) -> bool:
        raise NotImplementedError()

class SlidingWindowRateLimiter:
    def __init__(self, max_requests: int, window_seconds: float):
        # TODO: Implement sliding window
        raise NotImplementedError()

    def allow_request(self) -> bool:
        raise NotImplementedError()

File: java/src/main/java/com/learning/systemdesign/ex3RateLimiting.java
package com.learning.systemdesign;

import java.util.*;

public class ex3RateLimiting {
    public static class TokenBucketRateLimiter {
        public TokenBucketRateLimiter(int capacity, int refillRate) {
            // TODO: Implement token bucket algorithm
        }
        public boolean allowRequest() {
            throw new UnsupportedOperationException("Not implemented yet");
        }
    }

    public static class SlidingWindowRateLimiter {
        public SlidingWindowRateLimiter(int maxRequests, long windowMs) {
            // TODO: Implement sliding window
        }
        public boolean allowRequest() {
            throw new UnsupportedOperationException("Not implemented yet");
        }
    }
}
